feat(auth): Store refresh tokens in secure HTTP-only cookies

phanDreams opened the pull request:
This commit refactors the authentication system to enhance security by storing refresh tokens in HTTP-only cookies, mitigating XSS vulnerabilities.

Key changes include:
- Introduced a `CookieManager` service to abstract session management.
- Migrated cookie settings to a dedicated configuration file (`cookie_conf.go`) with production overrides.
- Updated `Login`, `OAuth`, and `RefreshToken` handlers to use the new cookie-based flow, removing the refresh token from API response bodies.
- Regenerated Swagger documentation to reflect the updated API contract.

This change simplifies client-side logic, centralizes session handling, and improves overall application security.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Server-side, cookie-based session management added; login, registration, and OAuth flows set session cookies (session_id, user_id, jti, refresh_token).
  - Token refresh now reads the refresh token from an HTTP-only cookie; no request body required.

- Documentation
  - API docs updated: POST /token/refresh no longer accepts a JSON body; GET /health no longer lists JSON consumption.

- Chores
  - Added production configuration with server, DB pooling, and cookie settings (production SameSite: lax).
  - Dependency upgrades and env-driven config/path overrides.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

phanDreams committed:
feat(auth): Store refresh tokens in secure HTTP-only cookies

This commit refactors the authentication system to enhance security by storing refresh tokens in HTTP-only cookies, mitigating XSS vulnerabilities.

Key changes include:
- Introduced a `CookieManager` service to abstract session management.
- Migrated cookie settings to a dedicated configuration file (`cookie_conf.go`) with production overrides.
- Updated `Login`, `OAuth`, and `RefreshToken` handlers to use the new cookie-based flow, removing the refresh token from API response bodies.
- Regenerated Swagger documentation to reflect the updated API contract.

This change simplifies client-side logic, centralizes session handling, and improves overall application security.

mbrovarska commented on internal/app/app.go:
os.LookupEnv doesn't return err, it means err is undefined and err.Error() can cause a compile error.

phanDreams replied:
@coderabbitai review

phanDreams replied:
Done!

phanDreams committed:
feat(auth): Transmit refresh token via secure cookie

This commit refactors the authentication flow to enhance security by using secure, HTTP-only cookies for refresh token transmission. Storing the refresh token in a cookie mitigates the risk of it being stolen via XSS attacks, as it is no longer accessible to client-side JavaScript.

Key changes include:

- **Cookie-based Token Storage**: The `Login`, `OAuth`, and `RefreshToken` handlers have been updated. Upon successful authentication or refresh, the new refresh token is now set in an HTTP-only cookie managed by the server.

- **New CookieManager Service**: A dedicated `CookieManager` service has been introduced to abstract all cookie-related operations, making the code cleaner and more maintainable.

- **Configuration**: Added a `cookie_config` section to the application configuration files (`config.yml`, `config_prod.yml`) to allow for environment-specific settings like `domain`, `secure`, `path`, and `SameSite` attributes.

- **Handler Logic Update**:
  - The `RefreshToken` handler now retrieves the refresh token directly from the incoming request's cookie, removing the need to pass it in the request body.
  - The `Login` and `OAuth` handlers now set the cookie after generating a new token pair.

The access token continues to be returned in the response body, as it is short-lived and intended for use in the `Authorization` header of subsequent API requests.

mbrovarska commented on internal/core/services/cookie_manager.go:
```diff
@@ -0,0 +1,134 @@
+	}
+	// Configure the default options for the session cookie.
+	store.Options(opt)
+
```
agree with coderabbit, CookieSecure=true

mbrovarska commented on internal/core/services/cookie_manager.go:
```diff
@@ -0,0 +1,134 @@
+func (cm *cookieManager) Set(c *gin.Context, key string, value interface{}) {
+	session := sessions.Default(c)
+	session.Set(key, value)
+}
```
func (cm *cookieManager) Set(c *gin.Context, key string, value interface{}) {
	session := sessions.Default(c)
	session.Set(key, value)
}
What do you return here?

mbrovarska commented on internal/core/services/cookie_manager.go:
```diff
@@ -0,0 +1,134 @@
+
+	// Save the session to apply the clearing and expiration.
+	return session.Save()
+}
```
agree

mbrovarska commented on internal/core/services/cookie_manager.go:
```diff
@@ -0,0 +1,134 @@
+
+	// Update default options
+	session.Options(cm.defaultOpts)
+}
```
what do you return here? Should it be return session.Save()?

mbrovarska commented on internal/core/services/cookie_manager.go:
```diff
@@ -0,0 +1,134 @@
+	for key, value := range values {
+		session.Set(key, value)
+	}
+}
```
agree

mbrovarska commented on internal/handlers/specialist_handler.go:
```diff
@@ -196,6 +211,32 @@ func (sh *SpecialistHandlerImpl) Login(c *gin.Context) {
+	}
+
+	sh.logger.Info("Session cookie set with",
+		zap.String("session_id", sessionID),
```
maybe you should not log sessionID? it can be sensitive information.

phanDreams committed:
refactor(auth): Migrate refresh token handling to secure HTTP-only cookies

Added improovements and fixed mistakes.
